src/free_addresses/free_addresses.go

```diff
@@ -0,0 +1,181 @@
+	return util.Execute(command, args...)
+}
+
+type status struct {
+	State  string `xml:"state,attr"`
+	Reason string `xml:"reason,attr"`
+}
+
+type address struct {
+	Addr     string `xml:"addr,attr"`
+	AddrType string `xml:"addrtype,attr"`
+}
+
+type host struct {
+	Status    status     `xml:"status"`
+	Addresses []*address `xml:"address"`
+}
+
+type nmaprun struct {
+	XMLName xml.Name `xml:"nmaprun"`
+	Hosts   []*host  `xml:"host"`
+}
+
+//  http://play.golang.org/p/m8TNTtygK0
+func inc(ip net.IP) {
+	for j := len(ip) - 1; j >= 0; j-- {
+		ip[j]++
+		if ip[j] > 0 {
+			break
+		}
+	}
+}
+
+func incSubnet(ip net.IP, mask int) {
+	if mask == 0 {
+		return
+	}
+	start := (mask - 1) / 8
+	add := byte(1 << (7 - ((mask - 1) % 8)))
+	prev := ip[start]
+	ip[start] += add
+	if ip[start] < prev {
+		for j := start - 1; j >= 0; j-- {
+			ip[j]++
+			if ip[j] > 0 {
+				break
+			}
+		}
+	}
+}
+
+func scanSubNetwork(subNetwork string, exe IExecute) ([]strfmt.IPv4, string, int) {
+	ip, cidr, err := net.ParseCIDR(subNetwork)
+	if err != nil {
+		wrapped := errors.Wrapf(err, "Network %s is not a valid CIDR", subNetwork)
+		logrus.WithError(wrapped).Warn("ParseCIDR")
+		return nil, wrapped.Error(), -1
+	}
+	if cidr.String() != subNetwork {
+		stderr := fmt.Sprintf("Requested CIDR %s is not equal to provided subNetwork %s", cidr.String(), subNetwork)
+		logrus.Warn(stderr)
+		return nil, stderr, -1
+	}
+	o, e, exitCode := exe.Execute("nmap", "-sn", "-n", "-oX", "-", "--max-retries", "0", subNetwork)
+	if exitCode != 0 {
+		logrus.Warnf("nmap failed with exit-code %d: %s", exitCode, e)
+		return nil, e, exitCode
+	}
+	var nmaprun nmaprun
+	err = xml.Unmarshal([]byte(o), &nmaprun)
+	if err != nil {
+		logrus.WithError(err).Warn("XML Unmarshal")
+		return nil, err.Error(), -1
+	}
+	occupiedAddesses := make(map[string]int)
+	for _, h := range nmaprun.Hosts {
+		if h.Status.State == "up" {
+			for _, a := range h.Addresses {
+				if a.AddrType == "ipv4" {
+					occupiedAddesses[a.Addr] = 0
+					break
+				}
+			}
+		}
+	}
+	ret := make([]strfmt.IPv4, 0)
+	for ip = ip.To4(); cidr.Contains(ip); inc(ip) {
+		_, ok := occupiedAddesses[ip.String()]
+		if !ok {
+			ret = append(ret, strfmt.IPv4(ip.String()))
+		}
+	}
+	return ret, "", 0
+}
+
+func scanNetwork(network string, exe IExecute) (*models.FreeNetworkAddresses, string, int) {
+	ret := models.FreeNetworkAddresses{
+		Network: network,
+	}
+	ip, cidr, err := net.ParseCIDR(network)
+	if err != nil {
+		wrapped := errors.Wrapf(err, "Network %s is not a valid CIDR", network)
+		logrus.WithError(wrapped).Warn("ParseCIDR")
+		return &ret, wrapped.Error(), -1
+	}
+	if cidr.String() != network {
+		stderr := fmt.Sprintf("Requested CIDR %s is not equal to provided network %s", cidr.String(), network)
+		logrus.Warn(stderr)
+		return &ret, stderr, -1
+	}
+	ones, _ := cidr.Mask.Size()
+
+	var mask int
+
+	if ones < MinSubnetMaskSize {
+		mask = MinSubnetMaskSize
+	} else {
+		mask = ones
+	}
```

Comment:
Is this a verification to make sure that the subnet range is not too big?

If so, better explain this.

Maybe put this in an aptly named validation function will even be better

Reply:
My decision was (and can be changed by setting the MinSubnetMaskSize), that each time nmap will receive subnets containing not more than 1024 addresses.  This corresponds to mask value 22.
If we want to change it, we need to change the constant value.
It is not a validation, but how the function divides the scans to segments.